crates/sui-framework/sources/kiosk.move

```diff
@@ -0,0 +1,249 @@
+        self: &mut Kiosk, cap: &Option<KioskOwnerCap>, item: T, ctx: &TxContext
+    ) {
+        try_access(self, cap, ctx);
+        dof::add(&mut self.id, Key { id: object::id(&item) }, item)
+    }
+
+    /// Take any object from the Safe.
+    /// Performs an authorization check to make sure only owner can do that.
+    public fun take<T: key + store>(
+        self: &mut Kiosk, id: ID, cap: &Option<KioskOwnerCap>, ctx: &TxContext
+    ): T {
+        try_access(self, cap, ctx);
+        dof::remove(&mut self.id, Key { id })
+    }
+
+    // === Trading functionality ===
+
+    /// Make an offer by setting a price for the item and making it publicly
+    /// purchasable by anyone on the network.
+    ///
+    /// Performs an authorization check to make sure only owner can sell.
+    public fun make_offer<T: key + store>(
+        self: &mut Kiosk, id: ID, price: u64, cap: &Option<KioskOwnerCap>, ctx: &TxContext
+    ) {
+        try_access(self, cap, ctx);
+        df::add(&mut self.id, Offer { id }, price)
+    }
+
+    /// Make a trade: pay the owner of the item.
+    ///
+    /// Received `TransferRequest` needs to be handled by the publisher of the T,
+    /// if they have a method implemented that allows a trade, it is possible to
+    /// request their approval (by calling some function) so that the trade can be
+    /// finalized.
+    ///
+    /// After a confirmation is received from the creator, an item can be placed to
+    /// a destination safe.
+    public fun purchase<T: key + store>(
+        self: &mut Kiosk, id: ID, payment: Coin<SUI>
+    ): TransferRequest<T> {
+        let price = df::remove<Offer, u64>(&mut self.id, Offer { id });
+        let inner = dof::remove<Key, T>(&mut self.id, Key { id });
+
+        assert!(price == coin::value(&payment), EIncorrectAmount);
+        balance::join(&mut self.profits, coin::into_balance(payment));
+
+        TransferRequest {
+            inner,
+            paid: price,
+            from: self.owner
+        }
+    }
+
+    /// Allow a `TransferRequest` to happen.
+    ///
+    /// Can only be performed by an `AcceptTransferCap` making the approval logic a resposibility
+    /// of the owner of the T.
+    /// It also means that unless there's a policy for T to allow transfers, Kiosk trades will not be possible.
+    public fun allow<T: key + store>(
+        _cap: &AllowTransferCap<T>, req: TransferRequest<T>
+    ): (ApprovedTransfer<T>, u64, Option<address>) {
+        let TransferRequest { inner, paid, from } = req;
+        (ApprovedTransfer { inner }, paid, from)
+    }
+
+    /// Finalize the trade by placing the purchased item into the destination Kiosk.
+    /// The operation can only be performed by the owner of the Kiosk be it an explicit
+    /// setting or an issued Capability.
+    public fun store_purchased<T: key + store>(
+        self: &mut Kiosk,
+        transfer: ApprovedTransfer<T>,
+        cap: &Option<KioskOwnerCap>,
+        ctx: &mut TxContext
+    ) {
```

Comment:
I think this is useful as a wrapper, but I would also expose a function that gives the caller the `T` directly (no recipient `Kiosk` required). The buyer might want to keep the object as a single-owner obj, use it in a game, wrap it, etc.